src/safeposix/filesystem.rs

```diff
@@ -117,16 +142,44 @@ pub fn load_fs_special_files() {
     }
 }
 
-pub fn persist_metadata() {
+// Serialize Metadata Struct to JSON, write to file
+pub fn persist_metadata(guard: Option<&FilesystemMetadata>) {
+    let ourreader;
+    //Acquire a readlock if we were not passed in a reference
+    let metadata = if let Some(rl) = guard {rl} else {
+        ourreader = FS_METADATA.read().unwrap(); 
+        &ourreader
+    };
+
+    // Serialize metadata to string
+    let metadatastring = interface::serde_serialize_to_string(&*metadata).unwrap();
+    
+    // remove file if it exists, assigning it to nothing to avoid the compiler yelling about unused result
+    let _ = interface::removefile(METADATAFILENAME.to_string());
+
+    // write to file
+    let mut metadatafo = interface::openfile(METADATAFILENAME.to_string(), true).unwrap();
+    metadatafo.writefile_from_string(metadatastring, 0).unwrap();
+    metadatafo.close().unwrap();
 }
 
-pub fn restore_metadata() {
+// Read file, and deserialize json to FS METADATA
+pub fn restore_metadata() -> FilesystemMetadata {
```

Comment:
I think the assymetry in persist_metadata and restore_metadata bears explaining in a comment. One directly uses the FSMETADATA value and the other does not. Maybe restore_metadata isn't the right name anymore and something like reconstruct_metadata would make that clearer as well?